### PR TITLE
Code review by Josh Pai

### PR DESCRIPTION
**Pull request by Josh-Pai**

### CODE REVIEW

**Comment by Josh-Pai** [comment]
## Three Technical Changes I Made:

1. **Asynchronous Code Improvements**
   - The code uses `.Result` to synchronously wait for asynchronous tasks. This is a common source of deadlocks and performance issues. Converted synchronous calls to asynchronous by using the `await` keyword and making relevant methods asynchronous.  
   *Ref*: [Asynchronous programming - C# | Microsoft Learn](https://learn.microsoft.com/en-us/dotnet/csharp/async)

2. **Replaced `Console.WriteLine()` with `_logger.LogError()`**
   - Improved logging consistency, centralized logging, and better testability.  
   *Ref*: [Logging in ASP.NET Core](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-6.0)

3. **Replaced `jsonString` to `breweryJson`**
   - Increased readability by using a more descriptive variable name.

4. **Added a Null Check to the Breweries List**
   - Added `breweries != null && breweries.Any()` to avoid situations where there is no data or to prevent null reference exceptions.

## Three Items I Learned from the Code Review:

1. **Using Git Branches Effectively**
   - Creating separate branches for new features or fixes helps keep the main codebase clean and organized, making collaboration easier.

2. **The Importance of Asynchronous Code**
   - Asynchronous code helps the application stay responsive, handling multiple tasks smoothly and efficiently.

3. **Enhancing Readability for Better Collaboration**
   - Refactoring code to improve readability makes it easier for others to understand and work on, reducing confusion and potential mistakes.


## Was the project available on GitHub on time?
  - Yes, it’s available on GitHub on time.

## Does the project compile?
  - Yes, it compiles successfully.

## Does the project have adequate documentation?
  - Yes, the code has adequate documentation. However, adding more comments inside the code files could further improve clarity.


## Three changes I committed to my group's repository in the first sprint

1. [Add the total record counts for the past month, 6 months, 1 year, and 2 years](https://github.com/Josh-Pai/IS7024/commit/231ece54720265f13396e1879c90c84b403fde7d)

2. [Add Feature to Query More Than 1000 Rows](https://github.com/Josh-Pai/IS7024/commit/561a3c4224abc2d964671fdc441bf544909d94f5)

3. [Add Total Nearby Crime Records by Latitude and Longitude](https://github.com/Josh-Pai/IS7024/commit/b1e785f9321be8176d141ae07639e84cabb1bf91)

File: Pages/Index.cshtml.cs [hunk 1]
```diff
@@ -38,57 +38,56 @@ private List<Brewery> GetBreweryData()
...
                 {
-                    breweries = Brewery.FromJson(jsonString);
+                    breweries = Brewery.FromJson(breweryJson);
                 }
                 else
                 {
                     foreach (string evt in validationEvents)
                     {
-                        Console.WriteLine(evt);
+                        _logger.LogError(evt);
```
Comment (Josh-Pai) [inline Pages/Index.cshtml.cs:59]: Improved logging consistency, centralized logging, and better testability

File: Pages/Index.cshtml.cs [hunk 2]
```diff
@@ -22,13 +22,13 @@ public IndexModel(ILogger<IndexModel> logger)
             _logger = logger;
         }
 
-        public void OnGet()
+        public async Task OnGetAsync()
```
Comment (Josh-Pai) [inline Pages/Index.cshtml.cs:25]: Change to asynchronous to prevent some performance issues

File: Pages/Index.cshtml.cs [hunk 3]
```diff
@@ -38,57 +38,56 @@ private List<Brewery> GetBreweryData()
...
-            IDictionary<string, string> stateAbbreviationMap = new Dictionary<string, string>();
-            foreach (var state in states)
-            {
-                stateAbbreviationMap[state.Name] = state.Abbreviation;
-            }
 
-            // Join brewery data with state abbreviation based on state name
-            foreach (var brewery in breweries)
+            HttpResponseMessage stateresult = await client.GetAsync("https://worldpopulationreview.com/static/states/abbr-name-list.json");
+            if (stateresult.IsSuccessStatusCode)
```
Comment (Josh-Pai) [inline Pages/Index.cshtml.cs:65]: Change to asynchronous to prevent some performance issues

File: Pages/Index.cshtml.cs [hunk 4]
```diff
@@ -38,57 +38,56 @@ private List<Brewery> GetBreweryData()
...
+            HttpResponseMessage result = await client.GetAsync("https://api.openbrewerydb.org/breweries");
             List<Brewery> breweries = new List<Brewery>();
 
             if (result.IsSuccessStatusCode)
             {
-                Task<string> readString = result.Content.ReadAsStringAsync();
-                string jsonString = readString.Result;
-                JSchema schema = JSchema.Parse(System.IO.File.ReadAllText("Data/breweryschema.json"));
-                JArray jsonarray = JArray.Parse(jsonString);
+                string breweryJson = await result.Content.ReadAsStringAsync();
```
Comment (Josh-Pai) [inline Pages/Index.cshtml.cs:46]: Replaced "jsonString" to "breweryJson" to increased readability

File: Pages/Index.cshtml [hunk 5]
```diff
@@ -26,6 +26,8 @@
                     <th>Street</th>
                 </tr>
             </thead>
+            @if (breweries != null && breweries.Any())
```
Comment (Josh-Pai) [inline Pages/Index.cshtml:29]: Added a Null Check to the Breweries List

Comment (discospiff) [reply Pages/Index.cshtml:29]: Good.

File: Pages/Index.cshtml.cs (hunk 2 above)
Comment (discospiff) [reply Pages/Index.cshtml.cs:25]: Good.

File: Pages/Index.cshtml.cs (hunk 4 above)
Comment (discospiff) [reply Pages/Index.cshtml.cs:46]: yes, good idea.

File: Pages/Index.cshtml.cs (hunk 1 above)
Comment (discospiff) [reply Pages/Index.cshtml.cs:59]: Reeeeeeeeeeally good idea.

**Comment by discospiff** [comment]
Several very good suggestions in this branch.
I recommend the team review each change and consider merging.

**Pull request merged**